src/core/evaluator.js

```diff
@@ -603,9 +601,9 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
           return new TranslatedFont('g_font_error',
             new ErrorFont('Type3 font load error: ' + reason), translated.font);
         });
-      }).then((translated) => {
+      }).then(async (translated) => {
         state.font = translated.font;
-        translated.send(this.handler);
+        await translated.send(this.handler);
```

Comment:
Does this mean the promise won't be resolved now until the font has been loaded on the main thread?

Reply:
Yes; but unfortunately I cannot see how we would avoid this, even if it makes font loading a *tiny* bit slower in general (given the need to wait here, rather than sending the font and hoping it works).

In order for the fallback to work the `disableFontFace` flag must have been set *before* any other operators are parsed, since otherwise [`PartialEvaluator.handleText`](https://github.com/mozilla/pdf.js/blob/c0d6e46e392b327996eb0964b7932cb5bdde1727/src/core/evaluator.js#L613) could easily be called before we've established that falling back is necessary. That would then, easily, lead to intermittent bugs where not all `FontPath`s are available to [`FontFaceObject.getPathGenerator`](https://github.com/mozilla/pdf.js/blob/c0d6e46e392b327996eb0964b7932cb5bdde1727/src/display/font_loader.js#L389) on the main-thread.

*In closing:* Given that all glyphs are now re-mapped to the PUAs, and that glyph rendering is thus *guaranteed* to break completely when a font is rejected by the sanitizer, in my opinion we don't have much choice here unless we'd somehow be able to generate font data that can never fail to load.

Reply:
An addendum: Please note that waiting for the main-thread to load certain data is already being done elsewhere on the worker-side. See for example the [`JpegStream` case](https://github.com/mozilla/pdf.js/blob/c0d6e46e392b327996eb0964b7932cb5bdde1727/src/core/evaluator.js#L438-L468), where we're checking that "simple" JPEG images can actually be rendered natively by the browser and otherwise fallback to the built-in https://github.com/mozilla/pdf.js/blob/master/src/core/jpg.js implementation.

Reply:
Just testing locally with tracemonkey this is adding 100-250ms to overall load time of page 1. I'd really like to avoid any new two way dependencies too.

How about we load the normal way as usual, but if on the main thread we find a font that error's during loading we re-request the page with a flag enabling your new async code? Or we could disable native fonts entirely for the entire page, if we want to avoid the back and forth.

Reply:
> Just testing locally with tracemonkey this is adding 100-250ms to overall load time of page 1. 

I noticed that too and have started to realize that the current solution, although simple, probably isn't going to cut it unfortunately.  I've been thinking of a slightly modified approach, to get rid of the waiting, and will see if that works out better.

> I'd really like to avoid any new two way dependencies too.

I understand what you mean, but I'm not sure if that can be avoided. In any case, the solution will not be as simply as I initially hoped.

> How about we load the normal way as usual, but if on the main thread we find a font that error's during loading we re-request the page with a flag enabling your new async code?

While this is a nice idea it seems like a can of worms I'd rather not open, nor have time to hack on, since there's *a lot* of parameters/state associated with parsing/rendering. Hence trying to cancel/restart everything correctly seems like it could become very "messy" to deal with.

>  Or we could disable native fonts entirely for the entire page, if we want to avoid the back and forth.

I'm not sure if that's what you meant here, but please note that you *really* don't want to unconditionally set `disableFontFace` for an entire page, since that will guarantee that non-embedded fonts cannot be rendered.

---

Anyway, I've got an idea for a modified approach which does away with the waiting and only will require *temporary* storing a bit more state on the worker-side. I'll see what I can come up with :-)

Reply:
I've closed this one, and opened PR #10539 instead, after having an out of the blue idea for a (hopefully) better solution.